src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8467,65 +8467,60 @@ private VariableComponentSyntax ParseDeconstructionComponent(bool topLevel = fal
             }
             else
             {
-                var type = ParseType(false);
+                TypeSyntax type;
+                bool reportMissingType = false;
+                if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken &&
+                    (this.PeekToken(1).Kind == SyntaxKind.CommaToken || this.PeekToken(1).Kind == SyntaxKind.CloseParenToken))
+                {
+                    var missingType = CreateMissingIdentifierToken();
+                    type = _syntaxFactory.IdentifierName(missingType);
+                    reportMissingType = true;
+                }
+                else
+                {
+                    type = ParseType(false);
+                }
+
                 VariableDesignationSyntax designation = ParseDeconstructionDesignation(topLevel);
+                if (reportMissingType)
+                {
+                    designation = this.AddError(designation, ErrorCode.ERR_TypeExpected);
+                }
+
                 result = _syntaxFactory.TypedVariableComponent(type, designation);
             }
 
             return
-                topLevel ? (DeconstructionComponentLooksGood(result) ? CheckFeatureAvailability(result, MessageID.IDS_FeatureTuples) : null) : result;
+                topLevel ? (CountTypesInComponent(result) > 0 ? CheckFeatureAvailability(result, MessageID.IDS_FeatureTuples) : null) : result;
         }
 
-        private bool DeconstructionComponentLooksGood(VariableComponentSyntax node, bool topLevel = true)
+        private int CountTypesInComponent(VariableComponentSyntax node)
```

Comment:
Consider returning a bool instead, since we only care about zero or non-zero. `ComponentHasTypes`?